daemons/ipa-kdb/ipa_kdb_principals.c

```diff
@@ -1630,7 +1656,24 @@ static krb5_error_code dbget_alias(krb5_context kcontext,
                                                  &trusted_realm);
     }
 
+    if (kerr == KRB5_KDB_NOENTRY) {
+        krb5_data *hstname = NULL;
+
+        if ((krb5_princ_size(kcontext, norm_princ) < 2) &&
+            ((flags & KRB5_KDB_FLAG_REFERRAL_OK) == 0)) {
```

Comment:
I suppose this test it supposed to catch enterprise principals with server referral enabled. wouldn't it be more meaningful to check the principal type directly instead of its number of components? Especially given the fact malformed enterprise principals are already rejected at line 1585.

Reply:
No, this is *not* for enterprise principals. This is for the case of cross-realm S4U2Proxy where a service would ask for a service ticket to the target service which is in a trusted domain. Kerberos client library will construct the target service principal as `name@OUR.REALM` regardless what is the actual realm of the target. KDC will trigger a search for this principal from KDB with a `KRB5_KDB_FLAG_REFERRAL_OK` flag set. We cannot rely on the parsed principal's realm (because that is ours). However, for server principal it would be the one with at least two components in the name and we need to look at the host name of the principal to understand to which trusted domain it belongs to.

Note that there might be more than two components in the server principal here because in Active Directory there are also 3-part SPNs (https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-adts/28ca4eca-0e0b-4666-9175-a37ccb8edada and https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-adts/cd328386-4d97-4666-be33-056545c1cad2). I added a logic to pick up the final component of the parsed principal which should be either a hostname or a domain name. In both cases we would get to compare that against our trusted domains' routing table to find the right domain.